src/middleware/inject_params.rs

```diff
@@ -6,74 +6,93 @@ use jsonrpsee::{
 use std::sync::Arc;
 
 use super::{Middleware, NextFn};
-use crate::{
-    api::{Api, ValueHandle},
-    config::MethodParam,
-    middleware::call::CallRequest,
-};
+use crate::{api::Api, config::MethodParam, middleware::call::CallRequest};
 
 pub enum InjectType {
     BlockHashAt(usize),
     BlockNumberAt(usize),
+    BlockTagAt(usize),
 }
 
 pub struct InjectParamsMiddleware {
-    head: ValueHandle<(JsonValue, u64)>,
+    api: Arc<Api>,
     inject: InjectType,
     params: Vec<MethodParam>,
 }
 
 impl InjectParamsMiddleware {
     pub fn new(api: Arc<Api>, inject: InjectType, params: Vec<MethodParam>) -> Self {
         Self {
-            head: api.get_head(),
+            api,
             inject,
             params,
         }
     }
 
-    fn get_index(&self) -> usize {
-        match self.inject {
-            InjectType::BlockHashAt(index) => index,
-            InjectType::BlockNumberAt(index) => index,
+    async fn replace_parameter(&self, request: &mut CallRequest) -> Option<(usize, JsonValue)> {
+        let (_, number) = self.api.get_head().read().await;
+        let finalized_head = self.api.finalized_head.borrow().clone();
+        let maybe_inject = match self.inject {
+            InjectType::BlockTagAt(index) => {
+                if let Some(param) = request.params.get(index).cloned() {
+                    if !param.is_string() {
+                        return None;
+                    }
+                    match param.as_str().unwrap_or_default() {
+                        "finalized" => {
```

Comment:
all the eth related code should be on a different file. don't mix substrate logic with eth logic